### ISSUE TITLE
`impl _ for Contract` semantics are weird

### ISSUE THREAD

**Issue by brockelmore** [FuelLabs/sway#2429]
Currently, when creating a contract, you define an abi and the implementation of that abi for `Contract`. This `Contract` type is some magical keyword struct that isn't imported and is seemingly expanded by the compiler.

Example:

https://github.com/FuelLabs/sway/blob/376010e7b8f84334ffacd0f99e64f185c971d213/examples/wallet_smart_contract/src/main.sw#L27

This is extremely unintuitive and just feels wrong as a programmer. I don't know what it expands into but even just having `Contract` be in the stdlib (and using any compiler magic you need to there) would be better. But if there are meaningful things that users of the lang may want later, consider maybe something like the following:

```rust
use std::Contract;

#[derive(Contract)]
struct MyContract; // empty struct in this example but may be worthwhile to think about what a non-empty type looks like/does?

abi Wallet {
    // snip
}

impl Wallet for MyContract {
    // snip
}
```

**Comment by brockelmore** [comment]
Yep i intuitively understood that abi == trait effectively. However, it feels weird to me to impl a trait on an unspecified type (i.e. is `impl abi MyAbi` feels more like a blanket trait impl than concrete for a specific contract).

Not sure I totally agree with the path set out personally, but `impl abi MyAbi` is better than the current situation. Defining the type in the manifest feels weird to me, and i lean towards more something like rust modules as a better path, i.e.:

```rust
// instead of `mod`
contract {
   // snip
}

library { 
    // snip
}

// ...
```

Forcing users to split files and define in a manifest feels not great imo

**Comment by sezna** [comment]
That make sense, We did originally have program types as wrappers with braces around the entire file, but found that the extra level of indentation it introduced to everything was undesirable. Especially considering most Sway files are either just a contract, a script, or a library, and few define a combination of these things. 

Rust does offer two syntax options: one specified by the file format layout, and one specified by wrapping something in `mod {}`. We could do something similar and allow for multiple syntaxes?

The Contract type may not be so unspecified. It could be expanded in the future to be used for type checking contract calls to foreign contracts with verified ABIs. Would it feel better to use the `Self` keyword instead of the `Contract` keyword, where `Self` refers to the encompassing contract? Or some other keyword?


```rust
impl MyAbi for Self {
 // snip
}
```


```rust
impl MyAbi for This{
 // snip
}
```



```rust
impl MyAbi for ThisContract {
 // snip
}
```

**Comment by brockelmore** [comment]
`Self` makes more sense (`This` isn't used in rust so would shy away from that). I think this may be ideal:

```rust
abi Wallet {
    // snip
}

 // effectively a named `mod`
contract MyContract {

    // internal impl maybe?
    impl Wallet for Self {
        // snip
    }
    
}

library MyLibrary {
    // snip
}
```

With an alternative being:
```rust
// file: MyContract.sw, in manifest have it be a `contract`

// maybe have an attribute at top as previously suggested just for clarity?
#[program_type(contract)]

abi Wallet {
    // snip
}

impl Wallet for Self {
    // snip
}
```

```rust
// file: MyLibrary.sw, in manifest have it be a `library`
#[program_type(library)]
fn /* ... snip ... */
```

**Comment by nfurfaro** [comment]
I like the combination of the `contract`  keword and the use of self as mentioned above:
```
 // effectively a named `mod`
contract MyContract {

    // internal impl maybe?
    impl Wallet for Self {
        // snip
    }
    
}
```
